radix-engine/src/engine/interpreters/scrypto_interpreter.rs

```diff
@@ -116,7 +116,8 @@ mod tests {
         // RFC 2056
         fn assert_all() {
             assert_send::<ScryptoInterpreter<DefaultWasmEngine>>();
-            assert_sync::<ScryptoInterpreter<DefaultWasmEngine>>();
+            // TODO: make sure engine is multi-thread safe!
```

Comment:
We create a new Radix engine for each transaction and single thread it - it's the ScryptoInterpreter caches we're sharing.

#FearlessConcurrency should mean this is fine? 👍

Reply:
As the runtime shares pointer to the cache, they may run into illegal state due to memory volatility. Like I said, we need dedicated story to make sure multi-thread work rather pretend to.

Reply:
I mean I've already checked this:

* The moka cache is thread safe
* The multiple copies of the engine share the WasmModule which is just data, and from which they instantiate separate / independent WasmInstances

Separately, the only other interaction between engine threads is in the database layer - where RocksDB is threadsafe. (and we ensure there are no commits during executing)

Is there anything I've missed?

Reply:
Mainly the `wasmer` implementation which stores a pointer which could be volatible.

Reply:
But that's the `WasmerInstance` which contains the pointer, not the `WasmerModule`. The `WasmerInstance` is not shared - it's created for each execution from the (shared) `WasmerModule` template.

Reply:
Indeed - Wasmer is written in Rust, and they have [good understanding of Send/Sync (It appears even Instance implements send!)](https://github.com/wasmerio/wasmer/pull/807), so we can trust that if Module is Sync then it is Sync. Indeed - I think it's essentially a data object :).

Reply:
You don't know what Rust is doing with the data under that pointer though since Rust doesn't know that it's a mutable object potentially being written to across multiple threads.

Reply:
I believe you're referring to the pointer that we create ourselves. This pointer isn't in the object which is being shared. This pointer is in the `WasmerInstance` (not shared) not the `WasmerModule` (shared).

This pointer in `WasmerInstance` has no affect on whether `WasmerModule` is `Sync`.

On a related note, I'm quite surprised that we use `    runtime_ptr: Arc<Mutex<usize>>,` and use `usize`, not a raw pointer type here, but anyway... I agree WasmerInstance possibly has an unsound impl of `Sync` (without looking into it further)... But feeling slightly like a broken record, we're not sharing `WasmerInstance`. And I've double-checked - the `WasmerInstance` pointer does not point into `WasmerModule` (which would be unsound for sure if it did). :)